cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesJarUtil.java

```diff
@@ -0,0 +1,238 @@
+        if (isDownloadSource) {
+            mavenUrl = mavenUrl + "&c=sources";
+        }
+
+        String fileName = "";
+
+        File[] jarFiles;
+
+        if (isDownloadSource) {
+            jarFiles = templatesDirectory.listFiles(fileNameFilterSources);
+        } else {
+            jarFiles = templatesDirectory.listFiles(fileNameFilterJar);
+        }
+
+        if (jarFiles.length <= 0 || isJarOutdated(jarFiles[0], mavenUrl, isDownloadSource, templatesDirectory)) {
+
+            HttpURLConnection conn = initializeConnection(mavenUrl);
+            InputStream inputStream = conn.getInputStream();
+            fileName = conn.getURL().getFile().substring(conn.getURL().getFile().lastIndexOf("/") + 1);
+            File file = new File(templatesDirectory.getPath() + File.separator + fileName);
+            Path targetPath = file.toPath();
+            if (!file.exists()) {
+                Files.copy(inputStream, targetPath, StandardCopyOption.REPLACE_EXISTING);
+            }
+            conn.disconnect();
+        } else {
+            fileName = jarFiles[0].getPath().substring(jarFiles[0].getPath().lastIndexOf(File.separator) + 1);
+
+        }
+        return fileName;
+    }
+
+    /**
+     * Checks whether there is a newer version of the templates on Maven
+     * @param jarFile
+     *            our jar file that we want to check
+     * @param mavenUrl
+     *            the URL from where we are going to retrieve the latest jar
+     * @param isDownloadSource
+     *            true if downloading source jar file
+     * @param templatesDirectory
+     *            directory where the templates jar are located
+     * @return true if our jar is outdated false otherwise
+     * @throws IOException
+     *             {@link IOException} occurred
+     * @throws ProtocolException
+     *             {@link ProtocolException} occurred
+     * @throws MalformedURLException
+     *             {@link MalformedURLException} occurred
+     */
+    private static boolean isJarOutdated(File jarFile, String mavenUrl, boolean isDownloadSource,
+        File templatesDirectory) throws MalformedURLException, ProtocolException, IOException {
+        String fileName = jarFile.getPath().substring(jarFile.getPath().lastIndexOf(File.separator) + 1);
+        Matcher m = matchJarVersion(fileName, isDownloadSource);
+        if (m.find() == false || m.group(2).isEmpty()) {
+            // Maybe the jar is corrupted, let's update it
+            return true;
+        } else {
+            // Split the version number because it contains dots e.g. 3.1.0
+            int[] versionNumbers = Arrays.stream(m.group(2).split("\\.")).mapToInt(Integer::parseInt).toArray();
+
+            // We do the same for the latest jar in Maven, therefore we need to download it
+            HttpURLConnection conn = initializeConnection(mavenUrl);
```

Comment:
could you please check, that HttpURLConnect might implement Closeable and thus should be in try( ... ) {} block to be automatically closed.

Reply:
ok,  just googled it. You are right, but we should do it with `conn.getInputStream()` definitively.

Reply:
Okay, done!